rust/sealevel/client/src/cmd_utils.rs

```diff
@@ -44,62 +47,91 @@ pub(crate) fn account_exists(client: &RpcClient, account: &Pubkey) -> Result<boo
+pub(crate) fn deploy_program(
     payer_keypair_path: &str,
-    program_keypair: &Keypair,
-    program_keypair_path: &str,
+    program_key_dir: &Path,
+    program_name: &str,
     program_path: &str,
     url: &str,
     local_domain: u32,
-) -> Result<(), ClientError> {
+) -> Result<Pubkey, ClientError> {
+    let (program_keypair, program_keypair_path) = create_or_get_keypair(
+        program_key_dir,
+        format!("{}-keypair.json", program_name).as_str(),
+    );
+    let program_id = program_keypair.pubkey();
+
     let client = RpcClient::new(url.to_string());
-    if !account_exists(&client, &program_keypair.pubkey())? {
-        deploy_program(
+    if account_exists(&client, &program_keypair.pubkey())? {
+        println!("Program {} already deployed", program_keypair.pubkey());
+        return Ok(program_id);
+    }
+
+    let (buffer_keypair, buffer_keypair_path) = create_or_get_keypair(
+        program_key_dir,
+        format!("{}-buffer.json", program_name).as_str(),
+    );
+
+    let mut compute_unit_price = get_compute_unit_price_micro_lamports_for_id(local_domain);
+
+    for attempt in 0..10 {
+        println!("Attempting program deploy Program ID: {}, buffer pubkey: {}, compute unit price: {}, attempt number {}", program_id, buffer_keypair.pubkey(), compute_unit_price, attempt);
+
+        if attempt > 0 {
+            println!(
+                "As this is not the first deploy attempt, the buffer {} is re-used",
+                buffer_keypair.pubkey()
+            );
+        }
+
+        let mut command = vec![
+            "solana",
+            "--url",
+            url,
+            "-k",
             payer_keypair_path,
-            program_keypair_path,
+            "program",
+            "deploy",
             program_path,
-            url,
-            local_domain,
+            "--upgrade-authority",
+            payer_keypair_path,
+            "--program-id",
+            program_keypair_path.to_str().unwrap(),
+            "--buffer",
+            buffer_keypair_path.to_str().unwrap(),
+        ];
+
+        let compute_unit_price_str = compute_unit_price.to_string();
+        command.extend(vec!["--with-compute-unit-price", &compute_unit_price_str]);
+
+        // Success!
+        if let Ok(true) = run_cmd(command.as_slice(), None, None) {
+            // TODO: use commitment level instead of just sleeping here?
+            println!("Sleeping for 2 seconds to fully allow program to be deployed");
+            sleep(Duration::from_secs(2));
+
+            return Ok(program_id);
+        }
+
+        println!(
+            "Failed to deploy program with compute unit price {}",
+            compute_unit_price
         );
-    } else {
-        println!("Program {} already deployed", program_keypair.pubkey());
-    }
 
-    Ok(())
-}
+        compute_unit_price = if compute_unit_price > 0 {
+            compute_unit_price * 11 / 10
+        } else {
+            1000
+        };
```

Comment:
would move this logic elsewhere and replace magic numbers with some consts or add comments

Reply:
added a comment